packages/roosterjs-react-common/lib/plugins/UndoWithImagePlugin.ts

```diff
@@ -0,0 +1,115 @@
+import { Undo } from 'roosterjs-editor-core';
+import { UndoSnapshotsService } from 'roosterjs-editor-core/lib/undo/UndoSnapshots';
+import { ImageManagerInteface, hasPlaceholder, UpdatePlaceholdersResult } from '../utils/ImageManager';
+
+// Max stack size that cannot be exceeded. When exceeded, old undo history will be dropped
+// to keep size under limit. This is kept at 10MB.
+const MAXSIZELIMIT = 10000000;
+
+export default class UndoWithImagePlugin extends Undo {
+    /**
+     * Create an instance of Undo
+     * @param preserveSnapshots True to preserve the snapshots after dispose, this allows
+     * this object to be reused when editor is disposed and created again
+     * @param bufferSize The buffer size for snapshots. Default value is 10MB, it is possible after
+     * placeholder to image resolution that buffer size is greater.
+     */
+    constructor(
+        private imageManager: ImageManagerInteface,
+        preserveSnapshots?: boolean,
+        private bufferSize: number = MAXSIZELIMIT
+    ) {
+        super(preserveSnapshots, bufferSize);
+    }
+
+    /**
+     * Dispose this plugin
+     */
+    public dispose(): void {
+        super.dispose();
+    }
+
+    protected getSnapshotsManager(): UndoSnapshotsService {
+        if (!this.undoSnapshots) {
+            this.undoSnapshots = new UndoSnapshotsWithImage(this.imageManager, this.bufferSize);
+        }
+
+        return this.undoSnapshots;
+    }
+}
+
+interface Snapshot {
+    value: string;
+    hasPlaceholder: boolean;
+}
+
+class UndoSnapshotsWithImage implements UndoSnapshotsService {
+    private snapshots: Snapshot[];
+    private totalSize: number;
+    private currentIndex: number;
+
+    constructor(private imageManager: ImageManagerInteface, private maxSize: number) {
+        this.snapshots = [];
+        this.totalSize = 0;
+        this.currentIndex = -1;
+    }
+
+    public canMove(delta: number): boolean {
+        const newIndex = this.currentIndex + delta;
+        return newIndex >= 0 && newIndex < this.snapshots.length;
+    }
+
+    public move(delta: number): string {
+        if (!this.canMove(delta)) {
+            return null;
+        }
+
+        this.currentIndex += delta;
+        const snapshot = this.snapshots[this.currentIndex];
+        if (snapshot.hasPlaceholder) {
+            const originalValue = snapshot.value;
+            const result: UpdatePlaceholdersResult = this.imageManager.updatePlaceholders(originalValue);
+            snapshot.hasPlaceholder = !result.resolvedAll;
+            snapshot.value = result.html;
+            const delta = originalValue.length - result.html.length;
+
+            // it is possible total size is greater at this point (unlikely if default spinner is used)
+            this.totalSize -= delta;
+        }
```

Comment:
Can you add a comment describing this logic for future references?

Reply:
yeah I can provide a high level detail, there isn't much code to grasp